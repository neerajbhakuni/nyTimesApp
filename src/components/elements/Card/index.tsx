import { useNavigate } from 'react-router-dom';

type CardProps = {
    card: {
        id: string
        url: string
        mediaUrl: string
        title: string
        abstract: string
        byline: string
        section: string
        publishedDate: string
    },
    index: number
};

const Card = ({ card, index }: CardProps) => {
    const { id, mediaUrl, title, abstract, byline, section, publishedDate } = card;
    const navigate = useNavigate();

    return (
        <div key={id} data-testid={`card-${index}`} onClick={() => navigate(`/detail/${id}`, { state: card })} className="mx-16 my-8 flex flex-col bg-white border border-gray-200 rounded-lg shadow md:flex-row md:max-w hover:bg-gray-100 dark:border-gray-700 dark:bg-gray-800 dark:hover:bg-gray-700" >
            <img data-testid={`card-image-${id}`} className="object-cover w-full rounded-t-lg h-96 md:h-auto md:w-48 md:rounded-none md:rounded-s-lg" src={mediaUrl} alt={`${title}`} />
            <div className="flex flex-col justify-between p-8 leading-normal">
                <p data-testid={`card-date-section-${id}`} className="mb-1 text-lg text-gray-700 dark:text-gray-400">
                    {`${publishedDate} | ${section}`}
                </p>
                <h5 data-testid={`card-title-${id}`} className="mb-2 text-3xl font-bold tracking-tight text-gray-900 dark:text-white">
                    {title}
                </h5>
                <p data-testid={`card-description-${id}`} className="mb-3 text-lg text-gray-700 dark:text-gray-400">
                    {abstract}
                </p>
                <p data-testid={`card-byline-${id}`} className="mb-2 text-lg text-gray-700 dark:text-gray-400">
                    {byline}
                </p>
            </div>
        </div >
    )
}

export default Card;
