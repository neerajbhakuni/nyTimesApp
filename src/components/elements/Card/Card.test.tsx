import { render, screen } from "@testing-library/react";
import fireEvent from '@testing-library/user-event';

import Card from "./index";

const mockUseNavigate = jest.fn();
jest.mock('react-router-dom', () => {
    return {
        ...jest.requireActual('react-router-dom') as any,
        useNavigate: () => mockUseNavigate
    }
});

const mockProps = {
    card: {
        id: "1",
        url: "https://test.com",
        mediaUrl: "https://test.com/image.png",
        title: "title",
        abstract: "abstract",
        byline: "by line",
        section: "section ",
        publishedDate: "",
    },
    index: 0,
};

describe("rendering Card screen", () => {
    test("Shows Card", () => {
        const { container } = render(<Card {...mockProps} />);
        expect(container).toMatchSnapshot();
    });

    test("should have card button", () => {
        render(<Card {...mockProps} />);
        const button = screen.getByTestId(`card-${mockProps?.index}`);
        fireEvent.click(button);
        expect(mockUseNavigate).toBeCalledTimes(1);
    });

    test("should have card title", () => {
        render(<Card {...mockProps} />);
        const parent = screen.getByTestId(`card-title-${mockProps?.card?.id}`);
        expect(parent).toBeInTheDocument();
    });

    test("should have card title text", () => {
        render(<Card {...mockProps} />);
        const parent = screen.getByText(`${mockProps?.card?.title}`);
        expect(parent).toBeInTheDocument();
    });

    test("should have card description", () => {
        render(<Card {...mockProps} />);
        const parent = screen.getByTestId(`card-description-${mockProps?.card?.id}`);
        expect(parent).toBeInTheDocument();
    });

    test("should have card description text", () => {
        render(<Card {...mockProps} />);
        const parent = screen.getByText(`${mockProps?.card?.abstract}`);
        expect(parent).toBeInTheDocument();
    });
});
