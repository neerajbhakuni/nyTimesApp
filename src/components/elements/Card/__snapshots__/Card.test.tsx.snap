// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering Card screen Shows Card 1`] = `
<div>
  <div
    class="mx-16 my-8 flex flex-col bg-white border border-gray-200 rounded-lg shadow md:flex-row md:max-w hover:bg-gray-100 dark:border-gray-700 dark:bg-gray-800 dark:hover:bg-gray-700"
    data-testid="card-0"
  >
    <img
      alt="title"
      class="object-cover w-full rounded-t-lg h-96 md:h-auto md:w-48 md:rounded-none md:rounded-s-lg	"
      data-testid="card-image-1"
      src="https://test.com/image.png"
    />
    <div
      class="flex flex-col justify-between p-8 leading-normal"
    >
      <p
        class="mb-1 text-lg text-gray-700 dark:text-gray-400"
        data-testid="card-date-section-1"
      >
         | section 
      </p>
      <h5
        class="mb-2 text-3xl font-bold tracking-tight text-gray-900 dark:text-white"
        data-testid="card-title-1"
      >
        title
      </h5>
      <p
        class="mb-3 text-lg text-gray-700 dark:text-gray-400"
        data-testid="card-description-1"
      >
        abstract
      </p>
      <p
        class="mb-2 text-lg text-gray-700 dark:text-gray-400"
        data-testid="card-byline-1"
      >
        by line
      </p>
    </div>
  </div>
</div>
`;
